www/js/copy.js

```diff
@@ -0,0 +1,184 @@
+
+// storing pulled shabad/bani from db in array to avoid calling db too much
+const pankteeArray = [];
+
+/**
+ *
+ * @param {String} htmlString explanation text from ceremonies
+ * uses sanitize html to remove all html tags from text
+ */
+function stripHTML(htmlString) {
+  const allowedTags = [];
+  const allowedAttributes = {};
+  const cleanText = sanitizeHtml(htmlString.English, { allowedTags, allowedAttributes });
+  return cleanText;
+}
+/**
+ *
+ * @param {Objects} unmapped rows from the DB
+ * remaps each row
+ */
+function remapShabad(unmapped) {
+  pankteeArray.length = 0;
+  unmapped.forEach(row => {
+    const remappedRow = global.controller.remapLine(row);
+    pankteeArray.push(remappedRow);
+  });
+}
+/**
+ *
+ * @param {Objects} unmapped rows fresh from DB
+ * for each row, it sees if row.Verse (normal panktees) row.Custom (custom symbols or text) is available
+ * if not, it just displays title (because those two will only not be available at the position of the title)
+ * Then, it remapps the correct part of the row Object
+ */
+function remapBani(unmapped) {
+  pankteeArray.length = 0;
+  for (let i = 0; i < unmapped.length; i += 1) {
+    const row = unmapped[i];
+    let toBeRemapped;
+    if (row.Verse) {
+      toBeRemapped = row.Verse;
+    } else if (row.Custom) {
+      toBeRemapped = row.Custom;
+    } else {
+      toBeRemapped = row.Bani.Gurmukhi;
+    }
+    const remapped = global.controller.remapLine(toBeRemapped);
+    pankteeArray.push(remapped);
+  }
+}
+/**
+ *
+ * @param {Objects} unmapped rows fresh from DB to be remapped
+ * similar to remapBani, except includes html stripping for explanation slides and looks for title at row.Ceremony
+ */
+function remapCeremony(unmapped) {
+  pankteeArray.length = 0;
+  let removeHTML;
+  for (let i = 0; i < unmapped.length; i += 1) {
+    const row = unmapped[i];
+    let toBeRemapped;
+    if (row.Verse) {
+      removeHTML = false;
+      toBeRemapped = row.Verse;
+    } else if (row.Custom) {
+      removeHTML = true;
+      toBeRemapped = row.Custom;
+    } else {
+      removeHTML = false;
+      toBeRemapped = row.Ceremony.Gurmukhi;
+    }
+    const remapped = global.controller.remapLine(toBeRemapped);
+    if (removeHTML) {
+      pankteeArray.push(stripHTML(remapped));
+    } else {
+      pankteeArray.push(remapped);
+    }
+  }
+  console.log(unmapped);
+}
+/**
+ *
+ * @param {number} id id given to the shabad/bani/ceremony in db
+ * @param {String} idType specifies if the gurbani is identified as a 'bani', 'shabad', or 'ceremony' in db
+ */
+async function loadFromDB(id, idType) {
+  let result;
+  if (idType === 'shabad') {
+    result = await baniDB.loadShabad(id);
+    remapShabad(result);
+  } else if (idType === 'bani') {
+    baniDB.loadBani(id, baniLengthMap[baniLength]).then(rows => {
+      remapBani(rows);
+    });
+  } else if (idType === 'ceremony') {
+    baniDB.loadCeremony(id).then(rows => {
```

Comment:
Ceremony isn't copying properly.

Reply:
As in the explanation slides or shabads?

Reply:
shabads.

Reply:
copying shabads in anand karaj working fine for me

Reply:
if you search for a shabad first, then load the ceremony and try to copy a line, it doesn't get the correct line.

Reply:
seems like the error is coming from the fact that `search.currentShabad.indexOf(search.currentLine);` isnt updating correctly;